lightningd/plugin.c

```diff
@@ -373,6 +407,182 @@ static bool plugin_opts_add(const struct plugin_request *req)
 	return true;
 }
 
+static void plugin_rpcmethod_destroy(struct json_command *cmd,
+				     struct jsonrpc *rpc)
+{
+	jsonrpc_command_remove(rpc, cmd->name);
+}
+
+static void plugin_rpcmethod_cb(const struct plugin_request *req,
+				struct plugin_rpc_request *rpc_req)
+{
+	struct json_stream *response;
+	const jsmntok_t *res;
+	assert(req->resulttok || req->errortok);
+
+	if (req->errortok) {
+		res = req->errortok;
+		command_fail(rpc_req->cmd, PLUGIN_ERROR, "%.*s",
+			     res->end - res->start, req->response + res->start);
+		tal_free(rpc_req);
+		return;
+	}
+
+	res = req->resulttok;
+	response = json_stream_success(rpc_req->cmd);
+
+	/* If this is a raw string result we'd be cropping the quotes from the
+	 * result, so add them back */
+	if (req->resulttok->type == JSMN_STRING)
+		json_add_member(response, NULL, "\"%.*s\"",
+				res->end - res->start,
+				req->response + res->start);
+	else
+		json_add_member(response, NULL, "%.*s", res->end - res->start,
+				req->response + res->start);
+
+	command_success(rpc_req->cmd, response);
+	tal_free(rpc_req);
+}
+
+static void plugin_rpcmethod_dispatch(struct command *cmd, const char *buffer,
+				      const jsmntok_t *params)
+{
+	const jsmntok_t *toks = params, *methtok, *idtok;
+	struct plugin_rpc_request *request;
+	struct plugins *plugins = cmd->ld->plugins;
+	struct plugin *plugin;
+
+	if (cmd->mode == CMD_USAGE) {
+		cmd->usage = "[params]";
+		return;
+	}
+
+	/* We're given the params, but we need to walk back to the
+	 * root object, so just walk backwards until the current
+	 * element has no parents, that's going to be the root
+	 * element. */
+	while (toks->parent != -1)
+		toks--;
+
+	methtok = json_get_member(buffer, toks, "method");
+	idtok = json_get_member(buffer, toks, "id");
+	/* We've parsed them before, these should not fail! */
+	assert(idtok != NULL && methtok != NULL);
+
+	request = tal(NULL, struct plugin_rpc_request);
+	request->method = tal_strndup(request, buffer + methtok->start,
+				      methtok->end - methtok->start);
+	request->id = tal_strndup(request, buffer + idtok->start,
+				      idtok->end - idtok->start);
+	request->params = tal_strndup(request, buffer + params->start,
+				      params->end - params->start);
+	request->plugin = NULL;
+	request->cmd = cmd;
+
+	/* Find the plugin that registered this RPC call */
+	list_for_each(&plugins->plugins, plugin, list) {
+		for (size_t i=0; i<tal_count(plugin->methods); i++) {
```

Comment:
If I'm reading this correctly, this means that every plugin will need to register independent method names, ie they should namespace their methods so there's no conflict with another plugin, otherwise you might call the wrong plugin! (probably the first one that registered that method name)

Reply:
That is a very nice catch. However I believe forcing users of an API to do something like name spacing correctly will probably be hard to enforce.
Therefor I believe that c-lightning should handle the name spacing and not the plugin developer. In that sense within clightning there would be a namespace given by something like `plugin_name::plugin->method`

Reply:
I think it's fair to push the "multiplexing" of multiple plugins wanting to register the same method out to a "dispatch plugin", i.e., a plugin that knows how to route them, and not handle them internally.